### PR TITLE
Oneclick unsubscribe

### PR DESCRIPTION
**Pull request by crabsinger**
## What
Create one click link to unsubscribe from receiving emails:
1- generate a token based on profile Id, email and email type
2- generate a link with the token and send it to post mark
3- adjust postmark template
4- create a landing page for the unsubscribe link

## Why

Allow users to easily unsubscribe from emails

## Screenshots (if appropriate):
image
image

## Before Merging
The production templates on postmark needs to be updated as well
1- A new layout needs to be updated "CoLinks with one click unsubscribe"
this layout will have the ubsubscribe link at the footer
image

2- CoLinks Unread Notifications template needs to use this new layout instead.

### CODE REVIEW

File: api/email/unsubscribe/[unsubscribeToken].ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import { decodeToken, isEmailType } from '../../../api-lib/email/unsubscribe';
+import { adminClient } from '../../../api-lib/gql/adminClient';
+import { NotFoundError, errorResponse } from '../../../api-lib/HttpError';
+
+export default async function handler(req: VercelRequest, res: VercelResponse) {
+  let unsubscribeToken: string | undefined;
+  if (typeof req.query.unsubscribeToken == 'string') {
+    unsubscribeToken = req.query.unsubscribeToken;
+  } else if (Array.isArray(req.query.unsubscribeToken)) {
```
Comment (crabsinger) [inline api/email/unsubscribe/[unsubscribeToken].ts:13]: do we need this? should never be an array, right?

File: api-lib/email/unsubscribe.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  const profileId = params.get('profileId');
+  const email = params.get('email');
+  const token = params.get('token');
+  const emailType = params.get('emailType');
+  if (!profileId || !email || !token || !emailType || !isEmailType(emailType)) {
+    throw new UnauthorizedError('Invalid unsubscribe token');
+  }
+
+  const generatedToken = genHmac(profileId, email, emailType);
+  if (token !== generatedToken) {
```
Comment (crabsinger) [inline api-lib/email/unsubscribe.ts:38]: use https://nodejs.org/dist/latest-v6.x/docs/api/crypto.html#crypto_crypto_timingsafeequal_a_b not ==

File: api/email/unsubscribe/[unsubscribeToken].ts (hunk 1 above)
Comment (sshmm) [reply api/email/unsubscribe/[unsubscribeToken].ts:13]: @crabsinger I think so but typescript is expecting either string or string[] so I handled it the same way api/cosoul/[address] is handled

File: api-lib/email/unsubscribe.ts (hunk 2 above)
Comment (sshmm) [reply api-lib/email/unsubscribe.ts]: @crabsinger  I am currently using timingSafeEqual. I converted the string into a buffer first.
I have tested the functionality after.

**Review by sshmm — COMMENTED** [review]
@crabsinger I think it's good to go

**Comment by sshmm** [comment]
@wingfeatherwave I think we may need a different background image for UnsubscribeEmailPage. I am using the same image as VerifyEmailPage

**Comment by sshmm** [comment]
@crabsinger  postmark needs to be updated first before merging

## Before Merging
The production templates on postmark needs to be updated as well
1- A new layout needs to be updated "CoLinks with one click unsubscribe"
this layout will have the ubsubscribe link at the footer
image

2- CoLinks Unread Notifications template needs to use this new layout instead.

**Review by sshmm — APPROVED** [review]

**Pull request merged**